extensions/iceberg/src/main/java/io/deephaven/iceberg/layout/IcebergFlatLayout.java

```diff
@@ -0,0 +1,136 @@
+                if (instructions.columnRenameMap() != null) {
+                    parquetInstructions = parquetInstructions.withColumnRenameMap(instructions.columnRenameMap());
+                }
+                if (instructions.s3Instructions().isPresent()) {
+                    parquetInstructions =
+                            parquetInstructions.withSpecialInstructions(instructions.s3Instructions().get());
+                }
+            }
+            return new IcebergTableParquetLocationKey(fileUri, 0, null, parquetInstructions);
+        }
+        throw new UnsupportedOperationException(String.format("%s:%d - an unsupported file format %s for URI '%s'",
+                table, snapshot.snapshotId(), format, fileUri));
+    }
+
+    /**
+     * @param table The {@link Table} to discover locations for.
+     * @param tableSnapshot The {@link Snapshot} from which to discover data files.
+     * @param fileIO The file IO to use for reading manifest data files.
+     * @param instructions The instructions for customizations while reading.
+     */
+    public IcebergFlatLayout(
+            @NotNull final Table table,
+            @NotNull final Snapshot tableSnapshot,
+            @NotNull final FileIO fileIO,
+            @NotNull final IcebergInstructions instructions) {
+        this.table = table;
+        this.snapshot = tableSnapshot;
+        this.fileIO = fileIO;
+        this.instructions = instructions;
+
+        this.cache = new HashMap<>();
+    }
+
+    public String toString() {
+        return IcebergFlatLayout.class.getSimpleName() + '[' + table.name() + ']';
+    }
+
+    @Override
+    public synchronized void findKeys(@NotNull final Consumer<IcebergTableLocationKey> locationKeyObserver) {
+        try {
+            // Retrieve the manifest files from the snapshot
+            final List<ManifestFile> manifestFiles = snapshot.allManifests(fileIO);
+            for (final ManifestFile manifestFile : manifestFiles) {
+                // Currently only can process manifest files with DATA content type.
+                if (manifestFile.content() != ManifestContent.DATA) {
+                    throw new TableDataException(
+                            String.format("%s:%d - only DATA manifest files are currently supported, encountered %s",
+                                    table, snapshot.snapshotId(), manifestFile.content()));
+                }
+                final ManifestReader<DataFile> reader = ManifestFiles.read(manifestFile, fileIO);
+                for (DataFile df : reader) {
+                    final URI fileUri = FileUtils.convertToURI(df.path().toString(), false);
+                    final IcebergTableLocationKey locationKey = cache.computeIfAbsent(fileUri, uri -> {
+                        final IcebergTableLocationKey key = locationKey(df.format(), fileUri);
+                        // Verify before caching.
+                        return key.verifyFileReader() ? key : null;
```

Comment:
You can skip the `verifyFileReader` part here, similar to how it we now skip it in `ParquetFlatPartitionedLayout` (after PR #5451) and have always been skipping it in `ParquetKeyValuePartitionedLayout`. 
`verifyFileReader` internally reads the metadata from the corresponding parquet file and we don't want to do that at the time of discovery but later when the user does a coalesce.